crates/module-system/module-implementations/sov-bank/src/query.rs

```diff
@@ -74,22 +36,31 @@ impl<S: Spec> Bank<S> {
     async fn route_gas_token_balance(
         state: ApiState<S, Self>,
         accessor: ApiStateAccessor<S>,
-        Path(user_address): Path<S::Address>,
+        Path(user_address): Path<String>,
     ) -> ApiResult<Coins> {
         Self::route_balance(state, accessor, Path((config_gas_token_id(), user_address))).await
     }
 
     async fn route_balance(
         state: ApiState<S, Self>,
         mut accessor: ApiStateAccessor<S>,
-        Path((token_id, user_address)): Path<(TokenId, S::Address)>,
+        Path((token_id, holder_str)): Path<(TokenId, String)>,
     ) -> ApiResult<Coins> {
-        let amount = state
-            .get_balance_of(&user_address, token_id, &mut accessor)
-            .unwrap_infallible()
-            .ok_or_else(|| errors::not_found_404("Balance", user_address))?;
-
-        Ok(Coins { amount, token_id }.into())
+        match TokenHolder::from_str(&holder_str) {
+            Ok(holder) => {
+                let amount = state
+                    .balances
+                    .get(&(holder.clone(), &token_id), &mut accessor)
+                    .unwrap_infallible()
+                    .ok_or_else(|| errors::not_found_404("Balance", holder))?;
+
+                Ok(Coins { amount, token_id }.into())
+            }
+            Err(err) => Err(errors::bad_request_400(
+                &format!("The given holder {holder_str} is not convertible to a TokenHolder"),
```

Comment:
Given that this will be the error provided as the API response, I would assume a majority of people querying it will be querying for a user balance and might not know about the TokenHolder type, maybe we can make it slightly more user-friendly:
```suggestion
                &format!("The given parameter {holder_str} is not a valid address or token holder"),
```
Thoughts?